quadrismegistus/prosodic
# Allow command-line use of prosodic

Hello!

### The issue
Previously, running `prosodic` would attempt to open the `bin/prosodic` file in a text editor on my Windows machine, rather than opening the interactive terminal.

### The fix
Now, by using setuptools' `entry_points`, the new function `start_interactive_mode` in `prosodic.py` is called whenever `prosodic` is used in the terminal. See https://stackoverflow.com/questions/774824/explain-python-entry-points for info on `entry_points`. The actual documentation might also be useful: https://setuptools.readthedocs.io/en/latest/userguide/entry_point.html, if you're curious.

This solution should be independent of file-system, operating system and python executable name (e.g. some people use `python38`, that might not have worked previously, after a quick glance of `bin/prosodic`)

The exact addition is:
```python
		entry_points = {
			'console_scripts': [
				'prosodic=prosodic.prosodic:start_interactive_mode'
			]
		},
```
The first `prosodic` is the name of the executable that can be run from the console. Then, `prosodic.prosodic` is the path (in terms of modules) to `prosodic/prosodic.py`, and `start_interactive_mode` is the required function that is called. Sadly it is not possible to simply run a file - we can only run a function. This meant that I had to modify `prosodic.py` slightly, to move the body of the interactive mode into a function.

Note that it gave an error if I didn't include `global lang`, which takes the `lang` variables defined previously. All other previously defined variables seemed to work correctly without needing to be included in global, (e.g. `config`, `languages`, etc.). It said that the local variable `lang` was not yet defined, and there's a chance that another variable will give this error, as I only quickly glanced for errors like these, and I didn't use every allowed command in the interactive mode, as I don't know the full functionality supported by the interactive mode.

### How to test
My recommendation is to clone this branch into a new folder, make a virtualenvironment for it, and perform `python setup.py install`. Afterwards, you should be able to run `prosodic` from any console where that virtualenvironment is activated, and will open the interactive terminal. 
(Note that after testing you can also install it on your "main" non-virtual-environment Python, which is used whenever you use `python ...` in a console that doesn't have a virtual environment activated, and then you should be able to use `prosodic` everywhere.)

### Output
```python
(.env38) PS C:\GitHub\prosodic> python setup.py install
...
Using c:\github\prosodic\.env38\lib\site-packages
Finished processing dependencies for prosodic==1.4.0
(.env38) PS C:\GitHub\prosodic> prosodic

################################################
## welcome to prosodic!                  v1.3 ##
################################################


        [please type a line of text, or enter one of the following commands:]
                /text   load a text
                /corpus load folder of texts
                /paste  enter multi-line text

                /mute   hide output from screen
                /exit   exit

>> [0.0s] prosodic:en$ This works, yay!
000001  this                    P:ðɪs                                   S:U     W:L
000002  works                   P:'wɛːks                                S:P     W:H
000003  yay                     P:'jeɪ                                  S:P     W:L

        [please type a line of text, or enter one of the following commands:]
                /text   load a text
                /corpus load folder of texts
                /paste  enter multi-line text

                /show   show annotations on input
                /tree   see phonological structure
                /query  query annotations

                /parse  parse metrically
                /meter  set the meter used for parsing
                /eval   evaluate this meter against a hand-tagged sample
                /maxent learn weights for meter using maxent

                /save   save previous output to file (except for /weight and /weight2; see /weightsave)
                /mute   hide output from screen
                /exit   exit

>> [4.43s] prosodic:en$ /exit

>> goodbye.
```

### Note
I only tested this on my one Windows machine, and haven't had the time to test it on my Ubuntu dualboot. I expect that it would work as I trust `setuptools` to be system-independent like it should be.

### What's next
Another thing to should be resolved, and I'll leave that up to you, @quadrismegistus, is that the interactive terminal shows `v1.3` as opposed to the real version. (Currently 1.4, but I would very much advise you to increment this version number and upload a new version to PyPI after looking at #29)

- Tom Aarsen
